library/core/src/krabcake.rs

```diff
@@ -0,0 +1,95 @@
+use core::arch::asm;
+
+mod vg_c_compatible {
+    const fn vg_userreq_tool_base(a: u64, b: u64) -> u64 {
+        ((a) & 0xff) << 24 | ((b) & 0xff) << 16
+    }
+
+    /// blah
+    #[allow(dead_code)] // For currently unused variants
+    #[derive(Debug, Clone, Copy)]
+    #[repr(u64)]
+    pub(super) enum ValgrindClientRequestCode {
```

Comment:
>embeds those values directly into the object code that gets generated

Not sure I quite follow what you mean in option 2; do you mean basically being able to have the MirPass insert the krabcake request into the MIR but without needing a Lang Item? ([example](https://github.com/bryangarza/rust/blob/66f9a3d6149651560bb1c520069811e856333019/compiler/rustc_mir_transform/src/valgrind_client_request.rs))

Or do you mean some method that doesn't use a MirPass?

---
>did you adopt this strategy because it was simpler to implement than option (2.)? Or am I missing some benefit of option (3.)?

The reason I made it a lang item is so that it would be more maintainable to have most of the logic defined as Rust code instead of needing to handwrite a lot of MIR. Also, I created a separate data-type `KrabcakeRequest` so that we don't couple the C enum to the enum that is used in `rustc`, since as you said, we could in the future change the format or approach that is used in `rustc` itself independently of what the request looks like when sent to Valgrind (or we could even use something other than/in addition to Valgrind)